app/gui/qt/mainwindow.cpp

```diff
@@ -2179,6 +2182,43 @@ void MainWindow::changeSystemPreAmp(int val, int silent)
     statusBar()->showMessage(tr("Updating System Volume..."), 2000);
 }
 
+// TODO: Implement real-time language switching
+void MainWindow::changeUILanguage(QString lang) {
+    if (lang != piSettings->language) {
+        std::cout << "Current language:  " << piSettings->language.toUtf8().constData() << std::endl;
+        std::cout << "New language selected: " << lang.toUtf8().constData() << std::endl;
+        QString old_lang = sonicPii18n->getNativeLanguageName(piSettings->language);
+        QString new_lang = sonicPii18n->getNativeLanguageName(lang);
+
+        // Load new language
+        //QString language = sonicPii18n->determineUILanguage(lang);
+        //sonicPii18n->loadTranslations(language);
+        //QString title_new = tr("Updated the UI language from %s to %s").arg();
+
+        QMessageBox msgBox(this);
+        msgBox.setText(QString(tr("You've selected a new language: %1")).arg(new_lang));
+        msgBox.setInformativeText(tr("Do you want to apply this language?\nApplying the new language will restart Sonic Pi."));
+        QPushButton *restartButton = msgBox.addButton(tr("Apply and Restart"), QMessageBox::ActionRole);
+        QPushButton *dismissButton = msgBox.addButton(tr("Cancel"), QMessageBox::RejectRole);
+        msgBox.setDefaultButton(restartButton);
+        msgBox.setIcon(QMessageBox::Information);
+        msgBox.exec();
+
```

Comment:
I've been trying to move code out of MainWindow; putting everything in this one file is a bit hard to follow.  You could, for example, put the handling of the messages and GUI in the settings panel.  It's OK for now though.

Reply:
I didn't even think of that :P That makes sense. I agree that having everything in mainwindow makes it hard to follow the code.
I've moved it to SettingsWidget :)